src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1447 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Semantics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundExpression : IExpression
+    {
+        ITypeSymbol IExpression.ResultType => this.Type;
+
+        OperationKind IOperation.Kind => this.ExpressionKind;
+       
+        object IExpression.ConstantValue => this.ConstantValue?.Value;
+
+        SyntaxNode IOperation.Syntax => this.Syntax;
+        
+        protected virtual OperationKind ExpressionKind => OperationKind.None;
+        // protected abstract OperationKind ExpressionKind { get; }
+    }
+
+    internal abstract partial class BoundNode : IOperationSearchable
+    {
+        IEnumerable<IOperation> IOperationSearchable.Descendants()
+        {
+            var list = new List<IOperation>();
+            new Collector(list).Visit(this);
+            list.RemoveAt(0);
+            return list;
+        }
+
+        IEnumerable<IOperation> IOperationSearchable.DescendantsAndSelf()
+        {
+            var list = new List<IOperation>();
+            new Collector(list).Visit(this);
+            return list;
+        }
+
+        private class Collector : BoundTreeWalker
+        {
+            private readonly List<IOperation> nodes;
+
+            public Collector(List<IOperation> nodes)
+            {
+                this.nodes = nodes;
+            }
+
+            public override BoundNode Visit(BoundNode node)
+            {
+                IOperation operation = node as IOperation;
+                if (operation != null)
+                {
+                    this.nodes.Add(operation);
+                }
+
+                return base.Visit(node);
+            }
+        }
+    }
+
+    partial class BoundCall : IInvocationExpression
+    {
+        IMethodSymbol IInvocationExpression.TargetMethod => this.Method;
+
+        IExpression IInvocationExpression.Instance => this.ReceiverOpt;
+
+        bool IInvocationExpression.IsVirtual => (this.Method.IsVirtual || this.Method.IsAbstract || this.Method.IsOverride) && !this.ReceiverOpt.SuppressVirtualCalls;
+
+        ImmutableArray<IArgument> IInvocationExpression.ArgumentsInSourceOrder
+        {
+            get
+            {
+                ImmutableArray<IArgument>.Builder sourceOrderArguments = ImmutableArray.CreateBuilder<IArgument>(this.Arguments.Length);
```

Comment:
OK. I don't understand the implications, but I'm happy to follow established practice.